### ISSUE TITLE
Make JsonPostProcessor (and perhaps others) public

### ISSUE THREAD

**Issue by Tetedeiench** [sawch/Patreon.Net#6]
Hi,

I've been using your work for interfacing with the mess that is Patreon's API, and it's been a lifesaver. Thanks !

I'm currently investigating Patreon's Webhooks. I'll gladly share my work with you later on (the small piece of code to verify a webhook's signature is especially valuable, despite being simple - it's the least I can do), and I was going to try and deserialize the body using your data structure and your deserialization algorithm.

Yet, as JsonPostProcessor is declared "internal", I can't do so, sadly :( I'll copy the class in my own project for now, but making that class public would allow us to parse data obtained from patreon in ways not yet supported by your project.

What is your opinion on this ?

Again, thanks for all this :)

**Comment by DurtyFree** [comment]
As this stuff is very much explicit to the Patreon API, I dont see the use cases for it to be publicly available.

What exactly is the use case you are implementing, which requires this stuff? Maybe it makes more sense to just implement those features as part of the Patreon.NET lib, as it might be needed by multiple people and is hardly patreon related anyway?

**Comment by Tetedeiench** [comment]
Well, i'm subscribing to Webhooks, and the body of the webhooks are really models you have already implemented and parsed. The pledge create webhook is sending you a serialized Member in its body.

By copying your JsonPostProcessor class into my solution, it enabled me to do this, for instance :
```
            // Verifying webhook signature and all before processing, ommitted for clarity purposes here
            var jsonSerializer = new JsonSerializer()
            {
                NullValueHandling = NullValueHandling.Ignore,
                MissingMemberHandling = MissingMemberHandling.Ignore
            };

            if (!JsonPostProcessor.Process(JObject.Parse(body), out JToken dataToken))
                throw new Exception("Unexpected or unparsable data");

            var member = dataToken.ToObject<Member>(jsonSerializer);
```

All in all, this changes enables us consumers to support raw data obtained from Patreon through any means that isn't implemented yet in the library, and benefit from the hard work you put into model parsing. Webhooks being the one i'm thinking of right now (I have no idea how to generalize webhook support in your library - you need the raw body and all to check the signature).

**Comment by DurtyFree** [comment]
As this webhook subscribing stuff is part of the Patreon API, I think it should be directly implemented and supported in this lib.

I also think in case Patreon changes or adds something, developers should better implement the new stuff or changes directly in the lib for everyone instead of doing their own thing using the internal stuff of the lib.

Just my 2 cents on the topic :D

**Comment by Tetedeiench** [comment]
Oh, I agree fully with you. But submitting your work to a public API in a lib like this tends to be scary (at least for me). Will my code be up to the standard this lib wants to stick to ? I submitted two small changes on public projects personally ( oxyplot and next-auth, the latter being adding an Oauth2 provider for Patreon actually ), and that was quite an emotional ride !

I was just suggesting a change that enhanced functionality at no cost for the lib, without resorting to what is, ultimately, a code smell (like me copying the class into my project).

I've got the signature validation nailed down, and subscribed to the one webhook event I need for my project. Once that's launched (as I need to prioritize this for... obvious reasons ;) ), i'll see if I can't generalize my work and submit it as a PR. Nervously !

**Comment by DurtyFree** [comment]
I can completely relate to your feelings. Thats what the complete Pull Request system exists for, the maintainer / contributors can review your code and suggest changes in case it doent fit the repositories coding styles or contains any bugs.

In my opinion changing those parts to be publicly available just breaks the overall design and purpose this lib exists for, its meant as a lib for consumers of the Patreon API and not people that want to use it as a base to basically implement patreon API stuff themselves. 
I guess it should never really be the case that people would want / need to implement anything from the Patreon API on their own because the lib doesnt contain it. This lib needs to be either feature complete or the devs that miss something need to open PRs to allow the changes to be implemented in the lib for everyone to use - as every dev that would use those very internal code parts would ultimatively just enhance the lib anyway in some way that others would profit from.

Basically that small change would just suggest wrong usage and behaviour (in terms of giving back to this public lib 😁)

**Comment by sawch** [comment]
Sorry for taking so long to see this! Yeah, ideally this library should handle everything in the Patreon V2 API. I initially wrote this because I needed it for a separate project and I unfortunately have more urgent things to do than to implement the rest of the endpoints like webhooks at the moment. I'm not heavily experienced with public collaboration as well so don't hesitate to throw your changes in here. Being worried of your own work not being up to snuff means you're at least intending to write good code so don't stress about it too much!

**Comment by sawch** [comment]
Will close this since it's been over a year without any discussion
